### ISSUE TITLE
Internal Compiler Error when compiling a BERT-based model with neuron-cc and TF 2.8

### ISSUE THREAD

**Issue by florianlaws** [aws-neuron/aws-neuron-sdk#623]
I'm trying to port a multi-task model based on BERT from Tensorflow 2.5 to 2.8, and get an Internal Compiler Error when 
compiling the model for Neuron:
```
Compiler status ERROR
WARNING:tensorflow:neuron-cc failed with:
2023-01-03 13:12:04.273542: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-01-03 13:12:04.284425: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-01-03 13:12:04.284918: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: ***************************************************************
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:  An Internal Compiler Error has occurred
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: ***************************************************************
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Error message:  ERROR: Failed in  /home/ubuntu/venvs/tf28/lib/python3.8/site-packages/neuroncc/starfish/bin/hlo2penguin --input /tmp/tmpgw4c_w25/hlo_module.pb --out-dir ./ --output penguin.py --disable-load-fusion
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Error class:    AssertionError
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Error location: Unknown
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Command line:   /home/ubuntu/venvs/tf28/bin/neuron-cc compile /tmp/tmpgw4c_w25/hlo_module.pb --framework XLA --pipeline compile SaveTemps --output /tmp/tmpgw4c_w25/hlo_module.neff --verbose=35 --fast-math=none --fp32-cast=matmult-fp16 --enable-fast-context-switch
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Internal details:
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/CommandDriver.py", line 224, in neuroncc.driver.CommandDriver.CommandDriver.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/commands/CompileCommand.py", line 576, in neuroncc.driver.commands.CompileCommand.CompileCommand.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/commands/CompileCommand.py", line 554, in neuroncc.driver.commands.CompileCommand.CompileCommand.runPipeline
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/commands/CompileCommand.py", line 558, in neuroncc.driver.commands.CompileCommand.CompileCommand.runPipeline
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Job.py", line 289, in neuroncc.driver.Job.SingleInputJob.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Pipeline.py", line 30, in neuroncc.driver.Pipeline.Pipeline.runSingleInput
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Job.py", line 289, in neuroncc.driver.Job.SingleInputJob.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Pipeline.py", line 30, in neuroncc.driver.Pipeline.Pipeline.runSingleInput
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Job.py", line 289, in neuroncc.driver.Job.SingleInputJob.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/jobs/Frontend.py", line 425, in neuroncc.driver.jobs.Frontend.Frontend.runSingleInput
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/jobs/Frontend.py", line 225, in neuroncc.driver.jobs.Frontend.Frontend.runXLAFrontend
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Version information:
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   Neuron Compiler version 1.13.5.0+7dcf000a6
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   HWM version 1.13.0.0-0
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   NEFF version Dynamic
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   TVM version 1.13.0.0+0
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   NumPy version 1.20.0
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   MXNet not available
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   TF not available
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Artifacts stored in: /tmp/tmpgw4c_w25

WARNING:tensorflow:neuron-cc failed with:
2023-01-03 13:12:04.273542: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-01-03 13:12:04.284425: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-01-03 13:12:04.284918: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:936] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: ***************************************************************
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:  An Internal Compiler Error has occurred
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: ***************************************************************
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Error message:  ERROR: Failed in  /home/ubuntu/venvs/tf28/lib/python3.8/site-packages/neuroncc/starfish/bin/hlo2penguin --input /tmp/tmpgw4c_w25/hlo_module.pb --out-dir ./ --output penguin.py --disable-load-fusion
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Error class:    AssertionError
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Error location: Unknown
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Command line:   /home/ubuntu/venvs/tf28/bin/neuron-cc compile /tmp/tmpgw4c_w25/hlo_module.pb --framework XLA --pipeline compile SaveTemps --output /tmp/tmpgw4c_w25/hlo_module.neff --verbose=35 --fast-math=none --fp32-cast=matmult-fp16 --enable-fast-context-switch
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Internal details:
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/CommandDriver.py", line 224, in neuroncc.driver.CommandDriver.CommandDriver.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/commands/CompileCommand.py", line 576, in neuroncc.driver.commands.CompileCommand.CompileCommand.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/commands/CompileCommand.py", line 554, in neuroncc.driver.commands.CompileCommand.CompileCommand.runPipeline
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/commands/CompileCommand.py", line 558, in neuroncc.driver.commands.CompileCommand.CompileCommand.runPipeline
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Job.py", line 289, in neuroncc.driver.Job.SingleInputJob.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Pipeline.py", line 30, in neuroncc.driver.Pipeline.Pipeline.runSingleInput
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Job.py", line 289, in neuroncc.driver.Job.SingleInputJob.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Pipeline.py", line 30, in neuroncc.driver.Pipeline.Pipeline.runSingleInput
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/Job.py", line 289, in neuroncc.driver.Job.SingleInputJob.run
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/jobs/Frontend.py", line 425, in neuroncc.driver.jobs.Frontend.Frontend.runSingleInput
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   File "neuroncc/driver/jobs/Frontend.py", line 225, in neuroncc.driver.jobs.Frontend.Frontend.runXLAFrontend
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Version information:
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   Neuron Compiler version 1.13.5.0+7dcf000a6
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   HWM version 1.13.0.0-0
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   NEFF version Dynamic
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   TVM version 1.13.0.0+0
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   NumPy version 1.20.0
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   MXNet not available
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]:   TF not available
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: 
01/03/2023 01:12:05 PM ERROR 29231 [neuron-cc]: Artifacts stored in: /tmp/tmpgw4c_w25

WARNING:tensorflow:
################### Attention!!! ###################
According to semantic analysis, this model requires
a neuron-cc 1.3 based fall-back code generator.
####################################################
WARNING:tensorflow:
################### Attention!!! ###################
According to semantic analysis, this model requires
a neuron-cc 1.3 based fall-back code generator.
####################################################
```

Even though neuron-cc states "TF not available", I have the following tensorflow packages installed
```
tensorflow                    2.8.0
tensorflow-addons             0.17.1
tensorflow-datasets           4.6.0
tensorflow-gpu                2.8.0
tensorflow-hub                0.12.0
tensorflow-io-gcs-filesystem  0.27.0
tensorflow-metadata           1.10.0
tensorflow-model-optimization 0.7.3
tensorflow-neuron             2.8.0.2.3.0.0
tensorflow-text               2.8.2
transformers                  4.23.1
```

Trying to use neuron-cc 1.3.7.0 instead (as suggested in the last lines of the error message) results in the following error message:
```
save_model_neuron: tracing...
Saving bert_regressor_classifier model in Neuron format failed. e: tfn.trace requires neuron-cc version >= 1.6.0.0; please update to latest neuron-cc by `pip install neuron-cc -U --extra-index-url=https://pip.repos.neuron.amazonaws.com`
```

It is difficult for me to provide a small reproducing example, but the model is specified roughly like this:
```
    transformer_layer = transformers.TFDistilBertModel.from_pretrained(transformer_version)

    input_word_ids = tf.keras.Input(shape=(None,), dtype=tf.int32, name="input_word_ids")
    input_matrix_regressor = tf.keras.Input(shape=(None,), dtype=tf.float32, name="input_matrix_regressor")
    input_matrix_classifier = tf.keras.Input(shape=(None,), dtype=tf.float32, name="input_matrix_classifier")

    x = transformer_layer.distilbert(input_word_ids)[0]
    bert_features = x[:, 0, :]
    x = tf.keras.layers.Dropout(dropout_rate_bert)(bert_features)

    x = tf.keras.layers.Dense(256,
                              kernel_regularizer=tf.keras.regularizers.l2(L2_hidden),
                              activation='elu',
                              name="dense_hidden")(x)

    x_regressor = tf.keras.layers.Dropout(dropout_rate_hidden)(x)
    output_regressor = tf.keras.layers.Dense(1,
                                             kernel_regularizer=tf.keras.regularizers.l2(L2_regressor),
                                             activation="relu",
                                             name=target_variables[0]+'_inner')(x_regressor)
    #output_inter_regressor = tf.math.multiply(output_inter_regressor, input_matrix_regressor, name="minutes_registered_mask")
    #output_regressor = tf.math.reduce_max(output_inter_regressor, axis=1, keepdims=False, name=target_variables[0])

    x_classifier = tf.keras.layers.Dropout(0.5)(x)
    output_inner_classifier = tf.keras.layers.Dense(num_companies,
                                              activation="sigmoid",
                                              name=target_variables[1]+'_inner')(x_classifier)
    output_inner_classifier = tf.math.multiply(output_inner_classifier, input_matrix_classifier, name="is_overrun_mask")
    output_classifier = tf.math.reduce_max(output_inner_classifier, axis=1, keepdims=True, name=target_variables[1])

    model = tf.keras.Model(inputs=[input_word_ids, input_matrix_regressor, input_matrix_classifier],
                           outputs=[output_regressor, output_classifier])
```

**Comment by jluntamazon** [comment]
We have reproduced an error with the additional layers and are looking into a fix. We are able to compile the distilbert model on its own, but the additional layers cause problems.

For completeness, would you be able to provide your input tensor sizes?

**Comment by florianlaws** [comment]
Hi @jluntamazon , thanks for your reply!
The input tensor sizes are as follows:
```
'input_word_ids': <tf.Tensor: shape=(1, 128), dtype=int32, ...>, 
'input_matrix_regressor': <tf.Tensor: shape=(1, 2), dtype=float32, ...>, 
'input_matrix_classifier': <tf.Tensor: shape=(1, 2), dtype=float32, ...>
```

Or, more generally, they have the sizes
```
'input_word_ids': <tf.Tensor: shape=(n, 128), dtype=int32, ...>, 
'input_matrix_regressor': <tf.Tensor: shape=(n, k), dtype=float32, ...>, 
'input_matrix_classifier': <tf.Tensor: shape=(n, k), dtype=float32, ...>
```
where `n` is the number of examples and `k` is a hyperparameter.
For compiling with neuron, however I use a single example, so `n=1`.

**Comment by awsilya** [comment]
We have done some more investigation regarding the behavior disparity between TF 2.5 and TF 2.8. It appears that on TF 2.5 compilation would not crash, but the model was not put onto neuron. The cause of the failure is the additional layers and not the distilbert itself. We recommend that you only compile the distilbert and run the rest of the model on CPU (example below). In the meantime we will work on our error reporting for tensorflow-neuron 2.

**Comment by florianlaws** [comment]
Hi @awsilya thanks for the message. I can't seem to find the example , could you please include the link?
Also, how would I verify if a model actually executes on Neuron? (or which parts of the model?)

**Comment by awsilya** [comment]
@florianlaws, 

`import transformers
import tensorflow as tf
import tensorflow.neuron as tfn


def build_model():

    dropout_rate_bert = 0.5
    dropout_rate_hidden = 0.5
    L2_hidden = 1.0
    L2_regressor = 1.0
    num_companies = 32

    #transformer_layer = transformers.TFDistilBertModel.from_pretrained('distilbert-base-uncased')

    #input_word_ids = tf.keras.Input(shape=(None,), dtype=tf.int32, name="input_word_ids")
    distilbert_out = tf.keras.Input(shape=(128, 768), dtype=tf.float32, name="distilbert_out")
    input_matrix_regressor = tf.keras.Input(shape=(32), dtype=tf.float32, name="input_matrix_regressor")
    input_matrix_classifier = tf.keras.Input(shape=(32), dtype=tf.float32, name="input_matrix_classifier")

    #x = transformer_layer.distilbert(input_word_ids)[0]
    #bert_features = x[:, 0, :]
    #distilbert_out = distilbert_out[0]
    bert_features = distilbert_out[:, 0, :]
    x = tf.keras.layers.Dropout(dropout_rate_bert)(bert_features)

    x = tf.keras.layers.Dense(256,
                              kernel_regularizer=tf.keras.regularizers.l2(L2_hidden),
                              activation='elu',
                              name="dense_hidden")(x)

    x_regressor = tf.keras.layers.Dropout(dropout_rate_hidden)(x)
    output_regressor = tf.keras.layers.Dense(1,
                                             kernel_regularizer=tf.keras.regularizers.l2(L2_regressor),
                                             activation="relu")(x_regressor)
    #output_inter_regressor = tf.math.multiply(output_inter_regressor, input_matrix_regressor, name="minutes_registered_mask")
    #output_regressor = tf.math.reduce_max(output_inter_regressor, axis=1, keepdims=False, name=target_variables[0])

    x_classifier = tf.keras.layers.Dropout(0.5)(x)
    output_inner_classifier = tf.keras.layers.Dense(num_companies,
                                              activation="sigmoid",
                                              )(x_classifier)
    output_inner_classifier = tf.math.multiply(output_inner_classifier, input_matrix_classifier, name="is_overrun_mask")
    output_classifier = tf.math.reduce_max(output_inner_classifier, axis=1, keepdims=True)

    model = tf.keras.Model(inputs=[distilbert_out, input_matrix_regressor, input_matrix_classifier],
                           outputs=[output_regressor, output_classifier])

    return model


if __name__ == '__main__':
    model = build_model()
    transformer_layer = transformers.TFDistilBertModel.from_pretrained('distilbert-base-uncased')
    model_neuron = tfn.trace(transformer_layer.distilbert, tf.random.uniform([1,128],maxval=100, dtype=tf.int32))

    for i in range(1000):
        distilbert_input = tf.random.uniform([1,128],maxval=100, dtype=tf.int32)
        distilbert_out = model_neuron(distilbert_input)
        example_inputs = [distilbert_out, tf.random.uniform([1,32]), tf.random.uniform([1,32])]
        print(model(example_inputs))
    `

**Comment by awsilya** [comment]
@florianlaws - attached.

[example.py.gz](https://github.com/aws-neuron/aws-neuron-sdk/files/10558950/example.py.gz)

**Comment by florianlaws** [comment]
Thanks for the example, I'll check it out!

**Comment by jeffhataws** [comment]
@florianlaws , please let us know if the example provided helped unblock you. Thanks!

**Comment by florianlaws** [comment]
@jeffhataws I think I understand the idea of it from the example (basically, splitting the model in two and running them in a pipeline). It will take me a while to try it in my production but, but I'll get back eventually.

**Comment by awsrjh** [comment]
Hi @florianlaws,

any update on your side ?

**Comment by mrnikwaws** [comment]
Hi @florianlaws,

I am closing this issue since sample code was supplied about 2 months back.  Please create a new issue if you are still having problems with your model.  I'd also recommend taking a look at running on inf2, since the hardware and model lowering work better in some cases using torch_neuronx.
